pkg/auth/oauth2authenticator.go

```diff
@@ -0,0 +1,252 @@
+	token, _ := getToken(config)
+	oauthConfig := getOAuthConfiguration(config)
+
+	return &oAuth2Authenticator{
+		httpClient:         httpClient,
+		config:             config,
+		oauthConfig:        oauthConfig,
+		token:              token,
+		openBrowserFunc:    openBrowser,
+		shutdownServerFunc: shutdownServer,
+	}
+}
+
+func (o *oAuth2Authenticator) persistToken(token *oauth2.Token) {
+	tokenstring, _ := json.Marshal(token)
+	o.config.Set(CONFIG_KEY_OAUTH_TOKEN, string(tokenstring))
+	o.token = token
+}
+
+func (o *oAuth2Authenticator) Authenticate() error {
+	var responseCode string
+	var responseState string
+	var responseError string
+	verifier := createVerifier(128)
+	state := string(createVerifier(15))
+	codeChallenge := getCodeChallenge(verifier)
+	ctx := context.Background()
+
+	if o.headless {
+		// TODO:
+	} else {
+		srv := &http.Server{}
+
+		http.HandleFunc(CALLBACK_PATH, func(w http.ResponseWriter, r *http.Request) {
+			responseError = html.EscapeString(r.URL.Query().Get("error"))
+			if len(responseError) > 0 {
+				details := html.EscapeString(r.URL.Query().Get("error_description"))
+				fmt.Fprintf(w, "Error during authentication! (%s)\n%s", responseError, details)
+			} else {
+				responseCode = html.EscapeString(r.URL.Query().Get("code"))
+				responseState = html.EscapeString(r.URL.Query().Get("state"))
+				fmt.Fprintf(w, "Succesfully Authenticated!")
+			}
+
+			go o.shutdownServerFunc(srv)
+		})
+
+		// iterate over different known ports if one fails
+		for _, port := range accepted_callback_ports {
+			srv.Addr = fmt.Sprintf("%s:%d", CALLBACK_HOSTNAME, port)
+			listener, err := net.Listen("tcp", srv.Addr)
+			if err != nil { // skip port if it can't be listened to
+				continue
+			}
+
+			// fill redirect url now that the port is known
+			o.oauthConfig.RedirectURL = getRedirectUri(port)
+
+			url := o.oauthConfig.AuthCodeURL(state, oauth2.AccessTypeOffline,
+				oauth2.SetAuthURLParam("code_challenge", codeChallenge),
+				oauth2.SetAuthURLParam("code_challenge_method", "s256"),
+				oauth2.SetAuthURLParam("response_type", "code"),
+				oauth2.SetAuthURLParam("scope", "offline_access"))
+
+			// launch browser
+			go o.openBrowserFunc(url)
+
+			err = srv.Serve(listener)
+			if err == http.ErrServerClosed { // if the server was shutdown normally, there is no need to iterate further
+				break
+			}
+		}
+	}
+
+	if len(responseError) > 0 {
+		return fmt.Errorf("authentication error: %s", responseError)
+	}
+
+	// check the response state before continuing
+	if state != responseState {
+		return fmt.Errorf("incorrect response state: %s != %s", responseState, state)
+	}
+
+	// Use the custom HTTP client when requesting a token.
+	if o.httpClient != nil {
+		ctx = context.WithValue(ctx, oauth2.HTTPClient, o.httpClient)
+	}
+
+	token, err := o.oauthConfig.Exchange(ctx, responseCode, oauth2.SetAuthURLParam("code_verifier", string(verifier)))
```

Comment:
Still not sure what the verifier does - is it a nonce/challenge? Would a UUID suffice here?

Reply:
The verifier and how it has to be generated is defined in the standard. It is based on a secret (called code) and it is used at the end of the flow by both parties to verify that the final request comes from the originating party.